core/store/src/lib.rs

```diff
@@ -175,25 +175,41 @@ impl StoreUpdate {
         StoreUpdate { storage, transaction, tries: Some(tries) }
     }
 
+    pub fn insert(&mut self, column: DBCol, key: &[u8], value: &[u8]) {
+        assert!(column.is_write_once());
+        self.transaction.insert(column, key.to_vec(), value.to_vec())
+    }
+
+    pub fn insert_ser<T: BorshSerialize>(
+        &mut self,
+        column: DBCol,
+        key: &[u8],
+        value: &T,
+    ) -> io::Result<()> {
+        let data = value.try_to_vec()?;
+        self.insert(column, key, &data);
+        Ok(())
+    }
+
     pub fn update_refcount(&mut self, column: DBCol, key: &[u8], value: &[u8], rc_delta: i64) {
         assert!(column.is_rc());
         let value = encode_value_with_rc(value, rc_delta);
         self.transaction.update_refcount(column, key.to_vec(), value)
     }
 
-    pub fn set(&mut self, column: DBCol, key: &[u8], value: &[u8]) {
-        assert!(!column.is_rc());
-        self.transaction.insert(column, key.to_vec(), value.to_vec())
+    pub fn update(&mut self, column: DBCol, key: &[u8], value: &[u8]) {
```

Comment:
Can’t say that I like the name of the method. I think it might be confusing. ‘update’ sounds like it can be used only when the value already exists. Similarly, ‘insert’ sounds like it can be used only when the value doesn’t exist.

Why do we actually need two separate methods? I don’t think it prevents any issues.

Reply:
maybe we should call it 'upsert' (insert or update?)

Reply:
I agree with @mina86 here that maybe having both insert and update are a little confusing. I guess the idea is that for write_once columns we should only use insert and for non write_once columns we should only use update. Maybe we can just use `update` for both. That would also make mark columns as write_once easier.

Reply:
Having two methods is intentional -- this gives visibility at the call-site what's the intended semantics of the operation. Basically, the end result I hope to achieve is that most code uses safer "insert", and, by looking for the usages of `.update`, we can quickly identify places which need extra careful handling. Even once we have strongly-typed access to the database, I think it'll still be beneficial to see the kind of access at the call-site. 


Though, indeed, we *can* use a single method, and do an assert near the db based on the column type.

Reply:
Changed the terminology to `set/insert`, in part to just minimize the diff. I am still not 100% in love with "insert", but I do think that having two methods here is valuable. 

@mzhangmzz, @mina86 did I convince you that separate APIs here are a good idea? Can revert to a single method if you still think that'd be better.

Reply:
Honestly I’m still not convinced.  I still don’t see why usages of `update` needs any extra careful handling.  What’s so much more dangerous about it compared to `insert` that we need to be able to identify where calls to `update` happen?

But for me the most important issue is with the names.  Like I’ve describe insert is operation where value does not exist and you insert a new one and update is an operation where value already exists and is changed.  This is just confusing that this PR uses those two terms with a completely different meaning.

Reply:
Sorry, missed that you’ve changed to set/insert in the end.  This is definitely a better naming and I probably could live with it.  It however leaves the question of why can’t I use set in places of insert.

Reply:
>What’s so much more dangerous about it compared to insert that we need to be able to identify where calls to update happen?

`.set` is racy -- if one thread does sets, and another does reads, the latter could potentially get inconsistent data. 

`.insert` is comparatively safer -- the reader thread might fail with error if the data is not written, but it won't siletnly read wrong data.

If you can live with it, I'll merge this, but I'll be open to unifying insert and set back, if we end up hating it in pratcice!